llvm/include/llvm/IR/RepoDefinition.h

```diff
@@ -107,56 +103,116 @@ struct GODigestState {
-    assert(isa<llvm::Function>(G) && "All contributions are functions!");
-    GOI[GO].Contributions.emplace_back(G);
+/// Helper Class for generating the hash for each global objects on the Module.
+/// Also caches hashes for previously generated global objects (GOs).
+class ModuleHashGenerator {
+
+  enum class Tags : char {
+    Backref = 'R',
+    End = 'E',
+    GO = 'T',
+  };
+
+  /// A map of GO to a unique number in the dependencies graph.
+  GOStateMap Visited;
+  /// Used to keep the GO information.
+  GOInfoMap GOIMap;
+  /// Memoize GO's hash if one of the following conditions is satisfied:
+  /// 1. GO is not in a cycle/loop.
+  /// 2. GO is in a self-loop but not if there is a loop involving another
+  /// object.
+  MemoizedHashes GOHashCache;
+  /// True if the module M is changed.
+  bool Changed = false;
+
+public:
+  ModuleHashGenerator() {}
+  /// calculate the digest for all global objects inside of the Module M.
+  void digestModule(Module &M);
+
+  /// Compute the GO's hash value using the digest of GO's dependencies if it is
+  /// stored in the RepoDefinition metadata.
+  /// \param GO The global object whose digest is to be computed.
+  /// \return The global object's digest value.
+  DigestType calculateDigestUseRepoDefinition(const GlobalObject *GO);
+
+  bool isChanged() const { return Changed; }
+
+  const GOInfoMap &getGOInfoMap() const { return GOIMap; }
+
+  const MemoizedHashes &getGOHashCache() const { return GOHashCache; }
+
+private:
+  /// Get global object hash value from memoized hashes.
+  /// \param GO The global object.
+  /// \return A tuple of global object's hash value and a bool which is true if
+  /// GO's hash is cached.
+  std::tuple<DigestType, bool> getGOHash(const GlobalObject *GO) const {
+    const auto Iter = GOHashCache.find_as(GO);
+    return (Iter != GOHashCache.end()) ? std::make_tuple(Iter->second, true)
+                                       : std::make_tuple(NullDigest, false);
   }
-  // Update G's dependencies.
-  GOInfo &GInfo = GOI[G];
-  GInfo.InitialDigest = std::move(Result.InitialDigest);
-  GInfo.Dependencies = std::move(Result.Dependencies);
-}
-
-// Create a global object information map and calculate the number of hashed
-// functions and variable  inside of the Module M.
-/// \param M Called module.
-/// \returns a pair containing a global object information map and a structure
-/// holding the number of hashed global variables and functions.
-std::pair<GOInfoMap, GONumber> calculateGONumAndGOIMap(Module &M);
+
+  /// Construct the global object information map (GOIMap) and calculate the
+  /// number of hashed functions and variables inside of the Module M.
+  void calculateGONumAndGOIMap(const Module &M);
+
+  /// Compute the hash value for the given global object GO.
+  /// \param GO The global object.
+  /// \param HasRepoDefinitionMD true if using the digest of GO's dependencies
+  ///        which is stored in the RepoDefinition metadata.
```

Comment:
You could delete "which is" here which would make this more direct.

Reply:
Removed, thanks.